src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1124,30 +1124,50 @@ protected override BoundNode VisitReturnStatementNoAdjust(BoundReturnStatement n
                 return null;
             }
 
-            if (_returnTypesOpt != null)
+            // Should not convert to method return type when inferring return type (when _returnTypesOpt != null).
+            if (_returnTypesOpt == null &&
+                TryGetReturnType(out TypeSymbolWithAnnotations returnType))
+            {
+                VisitOptionalImplicitConversion(expr, returnType, useLegacyWarnings: false, AssignmentKind.Return);
+            }
+            else
             {
-                // Inferring return type. Should not convert to method return type.
                 TypeSymbolWithAnnotations result = VisitRvalueWithResult(expr);
-                _returnTypesOpt.Add((node, result));
-                return null;
+                if (_returnTypesOpt != null)
+                {
+                    _returnTypesOpt.Add((node, result));
+                }
             }
 
-            TypeSymbolWithAnnotations returnType = GetReturnType();
-            VisitOptionalImplicitConversion(expr, returnType, useLegacyWarnings: false, AssignmentKind.Return);
             return null;
         }
 
-        private TypeSymbolWithAnnotations GetReturnType()
+        private bool TryGetReturnType(out TypeSymbolWithAnnotations type)
         {
             var method = (MethodSymbol)_symbol;
```

Comment:
When would this not be a `MethodSymbol`

Reply:
For `NullableWalker`, the `_symbol` field is always a `MethodSymbol`. But the field is defined in the base class and may be a `FieldSymbol` in other derived types.

---
In reply to: [260902151](https://github.com/dotnet/roslyn/pull/33719#discussion_r260902151) [](ancestors = 260902151)